packages/realm-react/src/useQuery.tsx

```diff
@@ -17,17 +17,80 @@
 ////////////////////////////////////////////////////////////////////////////
 
 import Realm from "realm";
-import { useEffect, useState } from "react";
+import { useEffect, useRef, useMemo, useReducer } from "react";
+
+const numericRegEx = /^-?\d+$/;
+
+export type UseQueryCollection<T> = Realm.Results<T & Realm.Object> & { version?: number };
 
 export function createUseQuery(useRealm: () => Realm) {
-  return function useQuery<T>(type: string | { new (): T }): Realm.Results<T> {
+  return function useQuery<T>(type: string | { new (): T }): UseQueryCollection<T> {
     const realm = useRealm();
-    const [collection, setCollection] = useState<Realm.Results<T & Realm.Object>>(() => realm.objects<T>(type));
+    const collectionCache = useRef(new Map());
+
+    // TODO: Since the listener doesn't return the objectId, we need to store it in reference to its index
+    // Refactor this when listeners return ObjectIds
+    const indexToIdMap = useRef<string[]>([]);
+
+    const collectionHandler = useMemo<ProxyHandler<Realm.Results<T & Realm.Object>>>(() => {
+      return {
+        get: function (target, key) {
+          // Pass functions through
+          const value = Reflect.get(target, key);
+          if (typeof value === "function") {
+            return value.bind(target);
+          }
+
+          // If the key is not numeric, pass it through
+          if (typeof key === "symbol" || !numericRegEx.test(key)) {
+            return value;
+          }
+
+          // If the key is numeric, check if we have a cached object for this key
+          const index = Number(key);
+          const object = target[index];
+          const id = object._objectId();
+
+          // If we do, return it...
+          if (collectionCache.current.get(id) && collectionCache.current.get(id).deref()) {
+            const collection = collectionCache.current.get(id).deref();
+            // if the colleciton was garbage collected, skip return and store the updated reference
+            if (collection) return collection;
+          }
+
+          // If not then this index has either not been accessed before, or has been invalidated due
+          // to a modification. Fetch it from the collection and store it in the cache
+          collectionCache.current.set(id, new WeakRef(object));
```

Comment:
Should we use a combination of table id and object id?

Reply:
I actually have a version where I'm using the schema name.  Is that accomplishing the same purpose?

Reply:
Yes, the name from the schema should be unique.

Reply:
The current implementation keeps the cache local to the `useQuery` function call.  This removes the need to have the table_id or type_name in the cache key.  the object_id will suffice.